File: shared/lib/warnings.py
#!/usr/bin/env python3
"""
shared/lib/warnings.py — structured warning helper for ai-squad hooks.

Exposes append_warning() for writing entries to
.agent-session/<task_id>/warnings.json in a schema-consistent,
atomic, append-only manner.

Schema (warnings.json):
  {
    "schema_version": 1,
    "warnings": [
      {
        "id": "<uuid4>",
        "timestamp": "<iso8601>",
        "source": "<caller>",
        "reason": "<short_snake_case>",
        "severity": "info|warning|error",
        "metadata": {<arbitrary>}
      }
    ]
  }

Security: task_id is validated against ^FEAT-\\d{3,4}$ before any file op.
PII must NOT be placed in metadata.
Atomic write: fcntl.LOCK_EX on the file during read-modify-write.
Pure stdlib. Python 3.8+.
"""

import fcntl
import json
import re
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

_TASK_ID_RE = re.compile(r"^(FEAT|DISC)-\d{3,4}$")
_VALID_SEVERITIES = {"info", "warning", "error"}
_SCHEMA_VERSION = 1


def _agent_session_root() -> Path:
    """Locate .agent-session/ relative to the repo root.

    Walks upward from this file's location looking for .agent-session/.
    Falls back to cwd if not found — callers pass task_id, not full path,
    so the root must be discoverable.
    """
    candidate = Path(__file__).resolve()
    for parent in candidate.parents:
        if (parent / ".agent-session").is_dir():
            return parent / ".agent-session"
    # fallback: use cwd
    return Path.cwd() / ".agent-session"


def append_warning(
    task_id: str,
    reason: str,
    source: str,
    metadata: dict[str, Any] | None = None,
    severity: str = "warning",
) -> dict[str, Any]:
    """Append a warning entry to .agent-session/<task_id>/warnings.json.

    Parameters
    ----------
    task_id:  Session identifier, e.g. 'FEAT-003'. Must match ^FEAT-\\d{3,4}$.
    reason:   Short snake_case reason string, e.g. 'empty_transcript'.
    source:   Emitter name, e.g. 'capture-pm-session' or 'verify-output-packet'.
    metadata: Optional dict of additional context. Must not contain PII.
    severity: One of 'info', 'warning', 'error'. Defaults to 'warning'.

    Returns
    -------
    The warning entry dict that was written.

    Raises
    ------
    ValueError: if task_id does not match ^FEAT-\\d{3,4}$ or severity is invalid.
    """
    if not _TASK_ID_RE.match(task_id):
        raise ValueError(
            f"Invalid task_id '{task_id}': must match ^FEAT-\\d{{3,4}}$"
        )
    if severity not in _VALID_SEVERITIES:
        raise ValueError(
            f"Invalid severity '{severity}': must be one of {sorted(_VALID_SEVERITIES)}"
        )

    warnings_path = _agent_session_root() / task_id / "warnings.json"
    warnings_path.parent.mkdir(parents=True, exist_ok=True)

    entry: dict[str, Any] = {
        "id": str(uuid.uuid4()),
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "source": source,
        "reason": reason,
        "severity": severity,
        "metadata": metadata if metadata is not None else {},
    }

    with warnings_path.open("a+", encoding="utf-8") as fh:
        fcntl.flock(fh.fileno(), fcntl.LOCK_EX)
        try:
            fh.seek(0)
            raw = fh.read().strip()
            if raw:
                try:
                    doc = json.loads(raw)
                except json.JSONDecodeError:
                    doc = {"schema_version": _SCHEMA_VERSION, "warnings": []}
            else:
                doc = {"schema_version": _SCHEMA_VERSION, "warnings": []}

            if not isinstance(doc, dict):
                doc = {"schema_version": _SCHEMA_VERSION, "warnings": []}
            if not isinstance(doc.get("warnings"), list):
                doc["warnings"] = []
            doc["schema_version"] = _SCHEMA_VERSION

            doc["warnings"].append(entry)

            fh.seek(0)
            fh.truncate()
            json.dump(doc, fh, indent=2)
            fh.write("\n")
        finally:
            fcntl.flock(fh.fileno(), fcntl.LOCK_UN)

    return entry

File: shared/lib/warnings.ts
/**
 * shared/lib/warnings.ts — structured warning helper for ai-squad hooks.
 *
 * Exposes appendWarning() for writing entries to
 * .agent-session/<task_id>/warnings.json in a schema-consistent,
 * atomic, append-only manner.
 *
 * Schema (warnings.json):
 *   {
 *     "schema_version": 1,
 *     "warnings": [
 *       {
 *         "id": "<uuid>",
 *         "timestamp": "<iso8601>",
 *         "source": "<caller>",
 *         "reason": "<short_snake_case>",
 *         "severity": "info|warning|error",
 *         "metadata": {<arbitrary>}
 *       }
 *     ]
 *   }
 *
 * Security: taskId is validated against /^(FEAT|DISC)-\d{3,4}$/ before any file op.
 * PII must NOT be placed in metadata.
 * Concurrency: lockfile via fs.openSync(wx) with retry-backoff (mirrors Python's fcntl.LOCK_EX).
 * Atomic write: tmp + rename pattern for cross-platform safety.
 * Pure Node.js stdlib (fs, crypto, path). No extra dependencies.
 */

import crypto from 'crypto';
import fs from 'fs';
import path from 'path';

// ---------------------------------------------------------------------------
// Simple lockfile helpers — wx-flag + retry-backoff (mirrors fcntl.LOCK_EX).
// Max total wait: ~100ms across 5 attempts (10ms, 20ms, 20ms, 25ms, 25ms).
// ---------------------------------------------------------------------------

const LOCK_MAX_ATTEMPTS = 5;
const LOCK_DELAYS_MS = [10, 20, 20, 25, 25];

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Acquire an exclusive lockfile at `lockPath`. Returns fd to release. */
async function acquireLock(lockPath: string): Promise<number> {
  for (let attempt = 0; attempt < LOCK_MAX_ATTEMPTS; attempt++) {
    try {
      // O_CREAT | O_EXCL | O_WRONLY — fails if file exists (atomic on POSIX + Windows NTFS)
      const fd = fs.openSync(lockPath, 'wx');
      return fd;
    } catch {
      if (attempt < LOCK_MAX_ATTEMPTS - 1) {
        await sleep(LOCK_DELAYS_MS[attempt] ?? 20);
      }
    }
  }
  throw new Error(`acquireLock: could not acquire ${lockPath} after ${LOCK_MAX_ATTEMPTS} attempts`);
}

/** Release a lockfile previously acquired via acquireLock. */
function releaseLock(fd: number, lockPath: string): void {
  try {
    fs.closeSync(fd);
  } catch {
    // ignore close errors
  }
  try {
    fs.unlinkSync(lockPath);
  } catch {
    // lock may already be gone — ignore
  }
}

const TASK_ID_RE = /^(FEAT|DISC)-\d{3,4}$/;
const VALID_SEVERITIES = new Set(['info', 'warning', 'error']);
const SCHEMA_VERSION = 1;

export interface Warning {
  id: string;
  timestamp: string;
  source: string;
  reason: string;
  severity: 'info' | 'warning' | 'error';
  metadata: Record<string, unknown>;
}

export interface AppendWarningOptions {
  reason: string;
  source: string;
  metadata?: Record<string, unknown>;
  severity?: 'info' | 'warning' | 'error';
}

interface WarningsDoc {
  schema_version: number;
  warnings: Warning[];
}

/**
 * Locate .agent-session/ by walking upward from this file's directory.
 * Falls back to process.cwd() + '.agent-session' if not found.
 */
function findAgentSessionRoot(): string {
  let dir = path.dirname(path.resolve(__filename));
  const maxDepth = 20;
  for (let i = 0; i < maxDepth; i++) {
    const candidate = path.join(dir, '.agent-session');
    if (fs.existsSync(candidate)) return candidate;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return path.join(process.cwd(), '.agent-session');
}

/**
 * Append a warning entry to .agent-session/<taskId>/warnings.json.
 *
 * Uses an exclusive lockfile (wx flag + retry-backoff) to prevent the
 * read-modify-write race when two concurrent callers both read before either
 * writes. Mirrors Python's fcntl.LOCK_EX semantics.
 *
 * @param taskId  Session identifier, e.g. 'FEAT-003'. Must match /^(FEAT|DISC)-\d{3,4}$/.
 * @param opts    Warning fields: reason, source, metadata (no PII), severity.
 * @returns       The warning entry that was written.
 * @throws        Error if taskId is invalid or severity is not in allowed set.
 */
export async function appendWarning(taskId: string, opts: AppendWarningOptions): Promise<Warning> {
  if (!TASK_ID_RE.test(taskId)) {
    throw new Error(`Invalid taskId '${taskId}': must match /^(FEAT|DISC)-\\d{3,4}$/`);
  }
  const severity = opts.severity ?? 'warning';
  if (!VALID_SEVERITIES.has(severity)) {
    throw new Error(`Invalid severity '${severity}': must be one of info, warning, error`);
  }

  const agentSessionRoot = findAgentSessionRoot();
  const taskDir = path.join(agentSessionRoot, taskId);
  const warningsPath = path.join(taskDir, 'warnings.json');
  const lockPath = warningsPath + '.lock';

  fs.mkdirSync(taskDir, { recursive: true });

  const entry: Warning = {
    id: crypto.randomUUID(),
    timestamp: new Date().toISOString(),
    source: opts.source,
    reason: opts.reason,
    severity,
    metadata: opts.metadata ?? {},
  };

  const lockFd = await acquireLock(lockPath);
  try {
    let doc: WarningsDoc = { schema_version: SCHEMA_VERSION, warnings: [] };
    if (fs.existsSync(warningsPath)) {
      try {
        const raw = fs.readFileSync(warningsPath, 'utf-8').trim();
        if (raw) {
          const parsed = JSON.parse(raw) as unknown;
          if (
            typeof parsed === 'object' &&
            parsed !== null &&
            !Array.isArray(parsed) &&
            Array.isArray((parsed as Record<string, unknown>).warnings)
          ) {
            doc = parsed as WarningsDoc;
          }
        }
      } catch {
        // malformed file — start fresh
        doc = { schema_version: SCHEMA_VERSION, warnings: [] };
      }
    }

    doc.schema_version = SCHEMA_VERSION;
    doc.warnings.push(entry);

    const tmp = warningsPath + '.tmp-' + crypto.randomBytes(4).toString('hex');
    fs.writeFileSync(tmp, JSON.stringify(doc, null, 2) + '\n', 'utf-8');
    fs.renameSync(tmp, warningsPath);
  } finally {
    releaseLock(lockFd, lockPath);
  }

  return entry;
}
